### PR TITLE
WIP: Add support for units in AffineTransform2D

### PR DESCRIPTION
**Pull request by Cadair**
This is rather rough at this point, but I am attempting to add support for but united and not-united evaluation of the `AffineTransform2D` model.

@astrofrog @nden 

_for some context this is me trying to get a full gWCS pipeline to support units, demonstration of this is here: https://gist.github.com/Cadair/a2506b1c8bdc611a29d6ab0d7a1b87a4_

### CODE REVIEW

File: astropy/modeling/projections.py [hunk 1]
```diff
@@ -1963,18 +1963,25 @@ def evaluate(cls, x, y, matrix, translation):
         x, y : array, float
               x and y coordinates
         """
+        if hasattr(x, 'unit') and hasattr(y, 'unit'):
```
Comment (Cadair) [inline astropy/modeling/projections.py]: I was trying to make this not quantity specific. (I wonder if there should be a unit ABC that all unit packages could use...)

File: astropy/modeling/projections.py [hunk 2]
```diff
@@ -1983,8 +1990,10 @@ def evaluate(cls, x, y, matrix, translation):
 
     @staticmethod
     def _create_augmented_matrix(matrix, translation):
+        if hasattr(matrix, 'unit') and matrix.unit and hasattr(translation, 'unit') and translation.unit:
+            translation = translation.to(matrix.unit)
```
Comment (Cadair) [inline astropy/modeling/projections.py]: I am not entirely sure this makes sense. I don't really know how the units in this matrix work.

**Comment by nden** [comment]
@cadair I view the affine transformation as a mathematical operation and in this sense I'm not sure parameters need to have units. Perhaps input units shoul dbe preserved as output units but this is as much as I would go.  I may be missing somethign though. In what context are you using it that you find it necessary to support units in parameters?

**Comment by Cadair** [comment]
@nden I am trying to build a WCS pipeline for a simple `HPLN/T-TAN` solar transform with units all the way through. What I have (link in original comment) is  the following pipeline:

`Shift | AffineTransformation2D | Pix2Sky_TAN | RotateNative2Celestial`

The units I expect through this pipeline (not parameters) is:

`pixel -> pixel | pixel -> degrees | degrees -> degrees | degrees -> degress`

This is with the `CDELT` scale being part of the affine transform.

Am I doing this wrong or missing a step or something?

**Comment by nden** [comment]
I use the affine transform in the same way (without units). Thinking more about this though, it isn't the right way to use it. Affine transformation represents a rotation, followed by translation. There's no change of units. In such transformation the output units should be the same as the inputs units --> the units of the `translation` parameter. Putting the scaling in the matrix is what brings the conversion of units and this is not how it's supposed to be used.

I haven't tried this but I think a better approach would be to take the scaling out as a separate transform and use th epixel_scale and plate_scale equivalences (because a scaling on its own should not introduce a change in units).  Just some thoughts ... no solution yet.

**Comment by nden** [comment]
Here's a kludgy way to do it, the scaling goes in the equivalences parameter.
```
scl = Scale(1*u.deg)
scl(3*u.pix, equivalences={'x': u.pixel_scale(2*u.deg/u.pix)})
<Quantity 6.0 deg>
```

**Comment by Cadair** [comment]
I mean, mathematically the affine can obviously do a scale if that is behaviour the model is going to support I feel it should do it in a way that preserves units?

I will experiment further with your alternative solution, which looks hopeful!

**Comment by nden** [comment]
@Cadair I am not suggesting using the `Scale` approach directly, I was just thinking aloud ...
I'm also looking into making Affine work.

**Comment by Cadair** [comment]
@nden I discovered #6822 in an attempt to use scale directly (which I do quite like from the perspective of making the steps of the transform easy to understand).

**Comment by pllim** [comment]
Can this be closed now that #6853 is merged?

**Pull request closed without merging**